conans/client/remote_manager.py

```diff
@@ -24,19 +24,62 @@ def upload_conan(self, conan_reference, remote):
         """Will upload the conans to the first remote"""
         basedir = self._client_cache.export(conan_reference)
         rel_files = self._client_cache.export_paths(conan_reference)
-
         the_files = build_files_set(basedir, rel_files)
-        the_files = compress_export_files(the_files)
+
+        # FIXME: Check modified exports by hand?
+
+        if EXPORT_TGZ_NAME not in the_files:
+            self._output.rewrite_line("Compressing exported files...")
+            the_files = compress_export_files(the_files)
+            export_tgz_path = os.path.join(basedir, EXPORT_TGZ_NAME)
+            save(export_tgz_path, the_files[EXPORT_TGZ_NAME])
+        else:
+            # Prepare files.
+            # FIXME: Should check if exits conanfile and manifest???
+            # If we should check it, we should check it in compress_export_files too
+            tmp = {EXPORT_TGZ_NAME: the_files[EXPORT_TGZ_NAME]["contents"]}
+            if CONANFILE in the_files:
+                tmp[CONANFILE] = the_files[CONANFILE]["contents"]
+            if CONAN_MANIFEST in the_files:
+                tmp[CONAN_MANIFEST] = the_files[CONAN_MANIFEST]["contents"]
+            the_files = tmp
         return self._call_remote(remote, "upload_conan", conan_reference, the_files)
 
     def upload_package(self, package_reference, remote):
         """Will upload the package to the first remote"""
         basedir = self._client_cache.package(package_reference)
         rel_files = self._client_cache.package_paths(package_reference)
-
         the_files = build_files_set(basedir, rel_files)
-        self._output.rewrite_line("Compressing package...")
-        the_files = compress_package_files(the_files)
+        tgz_path = os.path.join(basedir, PACKAGE_TGZ_NAME)
+
+        # If package has been modified remove tgz to regenerate it
+        # FIXME: Should we raise or let user upload what he want to?
+        read_manifest, expected_manifest = self._client_cache.package_manifests(package_reference)
+        if read_manifest is None or read_manifest.file_sums != expected_manifest.file_sums:
+            if PACKAGE_TGZ_NAME in the_files:
+                del the_files[PACKAGE_TGZ_NAME]
+                try:
+                    os.unlink(tgz_path)
+                except Exception:
+                    pass
+            raise Exception("Can't upload corrupted package '%s'" % str(package_reference))
+
+        # Check if conan_package.tgz is present
+        if PACKAGE_TGZ_NAME not in the_files:
```

Comment:
I'm doing some profiling and I think we should remove those checks. It takes more than a second with a 300MB package.